### ISSUE TITLE
for loops maintain the same block on iteration, which is referenced in any closures generated within

### ISSUE THREAD

**Issue by thehowl** [gnolang/gno#1135]
title might be confusing, but here's the example:

```go
package main

func main() {
        var fns []func()
        for _, v := range []int{1, 2, 3} {
                x := v*100 + v
                fns = append(fns, func() { println(x) })
        }
        for _, fn := range fns {
                fn()
        }
}
```

(the `x :=` declaration is used to demonstrate that this does not reference `v`, whereby the behaviour talked about here would [actually be correct in the current go specification](https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable), though [this is due to change in go1.22](https://github.com/golang/go/issues/60078))

go behaviour (expected):

```
101
202
303
```

gno behaviour:

```
303
303
303
```

**Comment by ltzmaxwell** [comment]
I have a fix in #1585, it works like this:
```go
package main

func main() {
	var fns []func()
	for _, v := range []int{1, 2, 3} {
		x := v*100 + v
		fns = append(fns, func() { println(x) })
	}
	for _, fn := range fns {
		fn()
	}
}

// Output:
// 101
// 202
// 303
```

**Comment by deelawn** [comment]
Putting this here for reference.  Here are the attempts to fix this that have not been merged: #1768, #1585, https://github.com/deelawn/gno/pull/3

This is the one currently in progress: https://github.com/gnolang/gno/pull/1780

**Comment by ltzmaxwell** [comment]
> Putting this here for reference. Here are the attempts to fix this that have not been merged: #1768, #1585, [deelawn#3](https://github.com/deelawn/gno/pull/3)
> 
> This is the one currently in progress: #1780

the latest attempt: #1818. #1780 has been proved unsuitable for all scenarios.

**Comment by Kouteki** [comment]
@ltzmaxwell do we close #1818 or do you want to iterate on it?

**Comment by ltzmaxwell** [comment]
> @ltzmaxwell do we close #1818 or do you want to iterate on it?

it's actually ready for review. thanks.

**Comment by ltzmaxwell** [comment]
close this as solved by #2429. Go1.22 loopvar is not supported at the moment, defer for future optimization. cc: @moul @thehowl @jaekwon for visibility.
